reps/2023-10-13-accelerator-support.md

```diff
@@ -0,0 +1,234 @@
+
+```python
+@ray.remote(num_gpus=1)
+def task():
+    ...
+
+@ray.remote(num_gpus=1, accelerator_type="A100")
+def task():
+    ...
+
+@ray.remote(num_gpus=1, accelerator_type="Intel-Max-1550")
+def task():
+    ...
+
+@ray.remote(resources={"TPU": 1})
+def task():
+    ...
+
+@ray.remote(resources={"TPU": 1}, accelerator_type="TPU-V4")
+def task():
+    ...
+```
+
+##### Alternatives
+Other alternative APIs considered:
+
+Similar to `num_gpus`, having a top level parameter for each accelerator family:
+
+```python
+@ray.remote(num_tpus=1)
+def task():
+    ...
+
+@ray.remote(num_tpus=1, accelerator_type="TPU-V4")
+def task():
+    ...
+```
+
+A single `ACCELERATOR` resource for the ability to schedule a task on any accelerator
+and an `accelerator_family` is also introduced to allow specifying the family of the accelerators to use:
+
+```python
+@ray.remote(resources={"ACCELERATOR": 1})
+def task()
+    ...
+
+@ray.remote(resources={"ACCELERATOR": 1}, accelerator_family="TPU")
+def task()
+    ...
+
+@ray.remote(resources={"ACCELERATOR": 1}, accelerator_type="TPU-V4")
+def task()
+    ...
+```
+
+Similar to the above but having a top level `num_accelerators` parameter:
+
+```python
+@ray.remote(num_accelerators=1)
+def task()
+    ...
+
+@ray.remote(num_accelerators=1, accelerator_family="TPU")
+def task()
+    ...
+
+@ray.remote(num_accelerators=1, accelerator_type="TPU-V4")
+def task()
+    ...
+```
+
+#### Implementation
+
+Core defines an `Accelerator` base class and each accelerator needs to implement a subclass (e.g. `TPUAccelerator`):
+
+```python
+@DeveloperAPI
+class Accelerator(ABC):
+    @staticmethod
+    @abstractmethod
+    def is_available() -> bool:
+        """Detect if the accelerator is available on this node."""
+
+    @staticmethod
+    @abstractmethod
+    def get_resource_name() -> str:
+        """Get the name of the resource representing this accelerator."""
+
+    @staticmethod
+    @abstractmethod
+    def get_visible_accelerator_ids_env_var() -> str:
+        """Get the env var that sets the ids of visible accelerators."""
+
+    @staticmethod
+    @abstractmethod
+    def get_num_accelerators() -> int:
+        """Get the total number of accelerators on this node."""
+
+    @staticmethod
+    @abstractmethod
+    def get_visible_accelerator_ids() -> Optional[List[str]]:
+        """Get the ids of accelerators visible to Ray."""
+
+    @staticmethod
+    @abstractmethod
+    def get_accelerator_type() -> Optional[str]:
+        """Get the type of the accelerator on this node.
+
+        The result should only be used when get_num_accelerators() > 0.
+
+        Return None if it's unknown.
+        """
+
+    @staticmethod
+    @abstractmethod
+    def validate_resource_request_quantity(quantity: float) -> Optional[str]:
```

Comment:
why's the return value optional?

Reply:
This returns the error message if validation fails. Return None if it's valid.

Reply:
oh... maybe return a `Tuple[bool, str]` with `(valid, msg)`?

Reply:
Changed.